test/infrastructure/docker/docker/kind_manager.go

```diff
@@ -195,16 +197,24 @@ func createNode(ctx context.Context, opts *nodeCreateOpts) (*types.Node, error)
 	}
 	containerConfig.ExposedPorts = ports
 
-	if usernsRemap(ctx) {
+	cli, err := types.GetDockerClient()
+	if err != nil {
+		return nil, errors.Wrap(err, "container client error")
+	}
+
+	if usernsRemap(ctx, cli) {
 		// We need this argument in order to make this command work
 		// in systems that have userns-remap enabled on the docker daemon
 		hostConfig.UsernsMode = "host"
 	}
 
-	cli, err := types.GetDockerClient()
+	// Make sure we have the image
+	reader, err := cli.ImagePull(ctx, opts.Image, dockerTypes.ImagePullOptions{})
 	if err != nil {
-		return nil, errors.Wrap(err, "container client error")
+		return nil, errors.Wrap(err, "image pull error")
 	}
+	_, _ = io.Copy(io.Discard, reader)
```

Comment:
@stmcginnis So the problem was that we have to read from the reader, otherwise the pull does not work?

Reply:
Apart from that, just a nit/question. Do we need what we're reading here? (I guess it's the response of the ImagePull call?)

Reply:
Yeah, I'm actually not too happy about this but and would love to have someone come along and call me an idiot and explain it better. :)

To give the background, without this line, things were working just fine on my local machine (macOS, kind with a local registry, either tilt up or setting up would clusterctl) but the CI job would always fail with the logs indicating the image wasn't available.

It worked without the explicit pull of the image for me locally, but adding that seems like a good step. Better to be explicit. Adding the ImagePull initially didn't make a difference in the CI job though.

Digging in to the logs, there were some signs that the network connection was being closed or something, then every retry after that would immediately be rejected. To help troubleshoot, I added some logging statements that among other things would get this response and emit it to stdout to try to see what was happening in the logs. But after doing that the CI jobs started passing.

I removed that debug logging and things started failing again. Other than some `fmt.Printf` statements, the only functional difference was reading out whatever was in that response. So to just get that cleared out and closed, I tried this approach of basically sending it to `/dev/null`. Just adding that did the trick and the image pull was happy.

I still can't explain exactly why (which I don't like) and this seems like a really odd step to have to take (which I don't like), but it made things happy. I would love to hear it if anyone has a better idea of what and why the reader would need to be drained for the pull operation to complete. Or why it would work fine locally but not in the CI environment.

If not, then I think I do need to at least add a comment here to explain why this is being done so no one comes along later and decides to "clean up" this line as something not needed.

Reply:
For the record, here is what the `reader` contains from the `cli.ImagePull` call:

```json
{"status":"Pulling from kindest/haproxy","id":"2.1.1-alpine"}
{"status":"Digest: sha256:f47e192f899b409ce16b5d81c24a9bf2c5154c055c7d0d5f753179839be046e5"}
{"status":"Status: Image is up to date for kindest/haproxy:2.1.1-alpine"}
```

Reply:
Thx for the extensive explanation :)

Just to clarify: What do you mean with "kind with a local registry"?
I'm only familiar with basic kind usage, do you use something like described here? https://kind.sigs.k8s.io/docs/user/local-registry/

I suspect that it will be locally reproducable when kind actually has to pull the image.

Reply:
Yep, that exactly.

I thought it might be the case that I already had the image pulled, so I did delete it to force it to need to be pulled down again. I still was not able to get the failure to happen locally. I'll try to clean things up again and see.

Reply:
I can confirm. I'm also testing it locally and it *just* works :)

Reply:
I'll play around a bit here: https://github.com/kubernetes-sigs/cluster-api/pull/4449/files
Just out of curiousity. Really interesting behaviour.

Reply:
I stepped through the imagepull func via debugger and the whole imagepull call is definitely completely done after this func. For the imagepull itself it should not matter what we're doing with the reader.

Just an interesting pitfall (I don't think it applies here): Prow does not test exactly the commit from the PR, it always tests the commit which would exist if the PR would get merged. But no idea if this explains it in any way (I assume nothing changed on the main branch which is relevant here)

Reply:
I did get it to fail locally. Instead of using kind with the local registry, I just spun up a "normal" kind cluster. I also deleted the locally pulled image. In this case, without the `io.Copy` call in there it did fail. So maybe also a timing thing.

Still no explanation why it fails, but at least I don't need to use the CI infrastructure to test it. :)

Reply:
Interesting. I also tried a regular kind cluster and removed the haproxy image with `docker rmi -f`. I'm using this code currently: https://github.com/kubernetes-sigs/cluster-api/blob/5c765efdb195a48e00cae987fed191fdd468052e/test/infrastructure/docker/docker/kind_manager.go#L211-L216

Anything obvious you're doing differently?

Reply:
Hm interesting data point. The image is not immediately available after the image pull call. It takes roughly 1-2 sec after the call returned until the image is available. 

So I can reproduce the error on container create this way. But it works after a retry (next reconcile). Also the image is of course visible via `docker images` vs what was happening in CI yesterday

Reply:
Nothing obvious. But if it worked for me pulling from a local registry but fails otherwise, that kind of feels like it might be a timing issue. Maybe?

Reply:
I would say in the first try it's probably close that the image is available after the pull before the container create. But we have retries. So my current local behaviour does not explain the issues in CI.

Reply:
I"ll try a round with 10s after the image pull: https://github.com/kubernetes-sigs/cluster-api/pull/4449/commits/d458fa0442c567f1eee9c23c3da4a73ae5314d3c

Reply:
We could add a check for the presence of the image, but that would seem overkill. It _should_ just retry and work if it's not there initially like you see. So no idea why the behavior in CI is different.e

Reply:
Yup. I also wouldn't add a retry with the current info. I think the error in CI is different compared to my local one. I'm not sure how you're executing CAPD right now. I did the following: (I'm using Ubuntu 20.04 locally)
* kind create
* tilt up
* downscale CAPD
* run CAPD from the IDE

So my CAPD controller is communicating directly with the docker socket and not from inside the kind cluster and the docker socket mount. But that shouldn't lead to different results.

Reply:
If we don't find the error, I would probably run this next week on our on-premise Prow. There I have full access and can just debug the running Prow job.

Reply:
Interesting. The sleep fixed it: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api/4449/pull-cluster-api-e2e-main/1380140653710675968

So the strange thing is. Why did the retry not fix it before

Reply:
A theory:
* we call ImagePull
* ImagePull returns but image is not available yet
* ContainerCreate is triggered and fails because image does not exist yet

Retry fails because the way we're doing the imagepull it's "stuck" somehow (maybe between Docker and containerd (`the stream is done or WriteHeader was already called`)) and then all our retries also fail. Maybe it's because our container create all fails and then we're (maybe) using the same context all the time? Not sure if we get a new context when reconcile is retried.

Reply:
This change broke it again: https://github.com/kubernetes-sigs/cluster-api/pull/4449/commits/97124a7ae1cdca13b3840758f63f20bd79c98d86

I think the problem is probably when we just close the reader immediately when the pull is still in progress.
I think either waiting via sleep or reading the response (which might be something like a synchronous image pull) solves it.

I'll do two more tests:
1. move reader.Close into defer. This should still fail, because the reader is closed to early
2. Add a sleep in the defer reader.Close. This should work again as the close is done after the pull has finished

Reply:
In one of my attempts I did try passing context.Background() into the ImagePull call to give it its own context. It didn't seem to make a difference.

Something seems to mess with the network session though. From the tests I did, ContainerCreate should pull the image if it's not present. So something is getting in a weird funky state when this fails where both the ImagePull doesn't complete and subsequent attempts to get the image are all aborted immediately.

Reply:
Yup. I think the immediate close breaks the in-progress pull.

Reply:
I guess that has always been the problem. So this could basically mean we only need the container create, but we have to wait until it's done. The wait can probably by implemented by reading the whole body, which probably takes a few seconds when the containercreate iniitally pulls the image.

Reply:
Hm probably doesn't even really break the in-progress pull in a way that it get's stuck. The close just cancels the call and we're probably doing this on each retry.

Reply:
Think I found a clue in the [http.Do](https://golang.org/pkg/net/http/#Client.Do) documentation. About the response that ultimately gets returned via the docker SDK:

```
If the Body is not both read to EOF and closed, the Client's underlying RoundTripper (typically Transport) may not be able to re-use a persistent TCP connection to the server for a subsequent "keep-alive" request.
```

It must use a persistent TCP connection. So by my interpretation, if we don't read to EOF but we call `reader.Close()`, that messes with something internally that will cause all subsequent calls to fail with the `the stream is done or WriteHeader was already called` message.

So how I have it now (with a comment on their for future readers) does appear to be how this is supposed to be handled.

Reply:
Could explain these failure messages from the logs on one of the failed attempts:

```
time="2021-04-06T17:10:09.034704575Z" level=info msg="Pull session cancelled"
time="2021-04-06T17:10:09.276696089Z" level=error msg="Not continuing with pull after error: context canceled"
```

Reply:
I tested it a bit with a standalone program / scratch file; https://gist.github.com/sbueringer/4904386f11294e933199f2f969aebd6a

Reading the response body is a synchronous pull and the Docker engine just writes the current progress until the pull is done and close the connection at the end

Reply:
Tried to do the same, but in my case the ContainerCreate just returns an error "no such image". This would suggest we have to do an explicit (synchronous)  ImagePull before

Reply:
So tl;dr from my experiments:
* we need an ImagePull
* we can make the ImagePull synchronous by reading the body until the server closes the connection

Reply:
Yep, I agree. Having a sleep might work some of the time, but if we have larger images or slower network speeds it has the risk of not being long enough. In other cases, it would add unnecessary delay. So the only safe approach is to drain the response stream and close after EOF.